src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.ThreadPoolValueTaskSource.cs

```diff
@@ -0,0 +1,187 @@
+            }
+
+            public ValueTaskSourceStatus GetStatus(short token) => _source.GetStatus(token);
+            public void OnCompleted(Action<object?> continuation, object? state, short token, ValueTaskSourceOnCompletedFlags flags) =>
+                _source.OnCompleted(continuation, state, token, flags);
+            int IValueTaskSource<int>.GetResult(short token) => (int) GetResultAndRelease(token);
+            long IValueTaskSource<long>.GetResult(short token) => GetResultAndRelease(token);
+
+            void IThreadPoolWorkItem.Execute()
+            {
+                Debug.Assert(_operation >= Operation.Read && _operation <= Operation.WriteGather);
+
+                long result = 0;
+                try {
+                    bool notifyWhenUnblocked = false;
+                    try
+                    {
+                        // This is the operation's last chance to be cancelled.
+                        if (_cancellationToken.IsCancellationRequested)
+                        {
+                            _source.SetException(new OperationCanceledException(_cancellationToken));
+                            return;
+                        }
+
+                        notifyWhenUnblocked = ThreadPool.NotifyThreadBlocked();
+                        switch (_operation)
+                        {
+                            case Operation.Read:
+                                Memory<byte> writableSingleSegment = Unsafe.As<ReadOnlyMemory<byte>, Memory<byte>>(ref _singleSegment);
+                                result = RandomAccess.ReadAtOffset(_fileHandle, writableSingleSegment.Span, _fileOffset);
+                                break;
+                            case Operation.Write:
+                                result = RandomAccess.WriteAtOffset(_fileHandle, _singleSegment.Span, _fileOffset);
+                                break;
+                            case Operation.ReadScatter:
+                                Debug.Assert(_multiSegmentCollection != null && _multiSegmentCollection is IReadOnlyList<Memory<byte>>);
+                                result = RandomAccess.ReadScatterAtOffset(_fileHandle, (IReadOnlyList<Memory<byte>>) _multiSegmentCollection, _fileOffset);
+                                break;
+                            case Operation.WriteGather:
+                                Debug.Assert(_multiSegmentCollection != null && _multiSegmentCollection is IReadOnlyList<ReadOnlyMemory<byte>>);
+                                result = RandomAccess.WriteGatherAtOffset(_fileHandle, (IReadOnlyList<ReadOnlyMemory<byte>>)_multiSegmentCollection, _fileOffset);
+                                break;
+                        }
+                    } finally
+                    {
+                        if (notifyWhenUnblocked)
+                            ThreadPool.NotifyThreadUnblocked();
+                        _operation = Operation.None;
+                        _fileOffset = 0;
```

Comment:
Do these two actually need to be reset?  The rest potentially hold onto resources that should be let go, but these don't.

Reply:
It was for consistency but indeed there is no actual reason. I removed them.

Reply:
Oh, `_operation` actually needs to be set, it makes `ValidateInvariants` fail.